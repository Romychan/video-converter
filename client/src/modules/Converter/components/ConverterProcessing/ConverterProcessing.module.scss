.container {
  height: 100vh;
  padding: 0rem 15px;
  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: center;
}

.icon {
  width: 75px;
  height: 75px;
}

.button {
  margin-top: 1rem;
}
