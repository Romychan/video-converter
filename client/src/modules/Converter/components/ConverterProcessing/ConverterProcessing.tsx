import { useNavigate } from 'react-router-dom';
import { QueryStatus } from '@reduxjs/toolkit/dist/query';

import { DEFAULT_PATH } from '@router/routes';
import { downloadVideo } from '../../utils/downloadVideo';

import { useActions } from '@hooks/useActions';
import { useTypedSelector } from '@hooks/useTypedSelector';
import { useSubscribeToProgressQuery } from '../../store/converterApi';

import { Button } from '@ui/Button/Button';
import { Icon } from '@ui/Icon/Icon';
import { Loader } from '@ui/Loader/Loader';
import { Status } from '@ui/Status/Status';

import styles from './ConverterProcessing.module.scss';

interface IConverterProcessingProps {
  type: QueryStatus;
  file: string;
  onCancel: (value: boolean) => void;
}

export const ConverterProcessing = ({
  type,
  file,
  onCancel,
}: IConverterProcessingProps) => {
  const navigate = useNavigate();
  const format = useTypedSelector((state) => state.converter.format);
  const { resetSettingsStore, resetVideoStore } = useActions();
  const { data: percent } = useSubscribeToProgressQuery();

  const handleResetState = () => {
    onCancel(true);

    resetSettingsStore();
    resetVideoStore();

    navigate(DEFAULT_PATH);
  };

  if (type === 'rejected') {
    return (
      <div className={styles.container}>
        <Status
          title="Произошла ошибка"
          text="Во время обработки вашего файла произошла ошибка"
        >
          <Icon name="error" className={styles.icon} />
        </Status>

        <Button
          state="primary"
          onClick={() => onCancel(false)}
          className={styles.button}
        >
          Вернуться к настройкам
        </Button>
      </div>
    );
  }

  if (type === 'fulfilled') {
    downloadVideo(file, `filename.${format}`);

    return (
      <div className={styles.container}>
        <Status
          title="Видео успешно обработано"
          text="Видео будет автоматически скачано на ваш компьютер"
        >
          <Icon name="success" className={styles.icon} />
        </Status>

        <Button
          state="primary"
          onClick={handleResetState}
          className={styles.button}
        >
          Загрузить новое видео
        </Button>
      </div>
    );
  }

  return (
    <div className={styles.container}>
      <Status
        title="Идет обработка видео"
        text="Этот процесс может занять некоторое время"
      >
        <Loader progress={Math.floor(percent || 0)} />
      </Status>
    </div>
  );
};
