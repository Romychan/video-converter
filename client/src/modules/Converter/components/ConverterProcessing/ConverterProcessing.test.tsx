import { screen } from '@testing-library/react';
import { vi } from 'vitest';

import { QueryStatus } from '@reduxjs/toolkit/dist/query';

import { renderTestApp } from '@tests/helpers/renderTestApp';
import { ConverterProcessing } from './ConverterProcessing';

describe('ConverterProcessing', () => {
  it('processing video work correctly', () => {
    const handleChange = vi.fn();

    renderTestApp(
      <ConverterProcessing
        type={QueryStatus.pending}
        file="test"
        onCancel={handleChange}
      />,
    );

    expect(screen.getByText(/Идет обработка видео/i)).toBeInTheDocument();
    expect(screen.getByText(/0%/i)).toBeInTheDocument();
  });

  it('error processing video work correctly', () => {
    const handleChange = vi.fn();

    renderTestApp(
      <ConverterProcessing
        type={QueryStatus.rejected}
        file="test"
        onCancel={handleChange}
      />,
    );

    expect(
      screen.getByText(/Во время обработки вашего файла произошла ошибка/i),
    ).toBeInTheDocument();
  });

  it('success processing video work correctly', () => {
    const handleChange = vi.fn();

    renderTestApp(
      <ConverterProcessing
        type={QueryStatus.fulfilled}
        file="test"
        onCancel={handleChange}
      />,
    );

    expect(screen.getByText(/Видео успешно обработано/i)).toBeInTheDocument();
  });

  it('should match snapshot', () => {
    const handleChange = vi.fn();
    const container = renderTestApp(
      <ConverterProcessing
        type={QueryStatus.pending}
        file="test"
        onCancel={handleChange}
      />,
    );

    expect(container).toMatchSnapshot();
  });
});
