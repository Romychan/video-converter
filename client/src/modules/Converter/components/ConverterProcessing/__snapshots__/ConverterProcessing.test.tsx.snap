// Vitest Snapshot v1

exports[`ConverterProcessing > should match snapshot 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="_container_46bb2b"
      >
        <div
          class="_container_e7c728"
        >
          <div
            class="_status_e7c728"
          >
            <div
              class="_container_d3c8b7"
            >
              <div
                class="_background_d3c8b7"
              >
                <div
                  class="_progress_d3c8b7"
                  data-testid="loader-progress"
                  style="width: 0%;"
                />
              </div>
              <span
                class="_label_d3c8b7 text text_small"
              >
                0%
              </span>
            </div>
          </div>
          <h1
            class="_title_e7c728 subtitle"
          >
            Идет обработка видео
          </h1>
          <p
            class="_text_e7c728 text"
            data-testid="status-text"
          >
            Этот процесс может занять некоторое время
          </p>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="_container_46bb2b"
    >
      <div
        class="_container_e7c728"
      >
        <div
          class="_status_e7c728"
        >
          <div
            class="_container_d3c8b7"
          >
            <div
              class="_background_d3c8b7"
            >
              <div
                class="_progress_d3c8b7"
                data-testid="loader-progress"
                style="width: 0%;"
              />
            </div>
            <span
              class="_label_d3c8b7 text text_small"
            >
              0%
            </span>
          </div>
        </div>
        <h1
          class="_title_e7c728 subtitle"
        >
          Идет обработка видео
        </h1>
        <p
          class="_text_e7c728 text"
          data-testid="status-text"
        >
          Этот процесс может занять некоторое время
        </p>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "store": {
    "@@observable": [Function],
    "dispatch": [Function],
    "getState": [Function],
    "replaceReducer": [Function],
    "subscribe": [Function],
  },
  "unmount": [Function],
}
`;
